### PR TITLE
Dev/preprocessrxn cpp

### PR DESCRIPTION
**Pull request by bp-kelley**
Implements #732

### CODE REVIEW

**Comment by bp-kelley** [comment]
Interesting note, doctests can't find Boost.Python.Functions because while they are labelled as "callable" they aren't labelled as "functions" so doctests aren't found in their docs.  I have a hacky solution I might try do clean up and submit to boost,

**Comment by greglandrum** [comment]
how about just having the documentation+doctest in the python module that imports the C++ module and attaching it to a lambda function?

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 1]
```diff
@@ -60,12 +60,12 @@ def PreprocessReaction(*a, **kw):
...
   [x.lstrip() for x in rdChemReactions.PreprocessReaction.__doc__.split("\n")])
-exec_(code,TestPreprocess.__dict__)
+exec_c(code,TestPreprocess.__dict__)
 
 def load_tests(loader, tests, ignore):
   finder=doctest.DocTestFinder(True)
-  #finder.find(Enumerator)
-  #tests.addTests(doctest.DocTestSuite(Enumerator))
+  finder.find(Enumerator)
+  tests.addTests(doctest.DocTestSuite(Enumerator))
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: This does the lambda wrapping thing, forgot to turn it back on.

**Comment by bp-kelley** [comment]
I did have a version of this, but forgot that I had turned it off (with most of the other tests, I really need an git_hook to protect me from this, this is the second time I've forgot to turn them back on:) )

I think it is a good idea in general, however I still think

```
 tests.addTests(BoostDocTestSuite(rdChemReactions))
```

Would be quite useful.  Interestingly, anything in a class or class method is found, it is just raw functions that have the issue.

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 2]
```diff
@@ -31,23 +31,42 @@
...
 from rdkit import RDConfig
-
+from rdkit.Chem.SimpleEnum import Enumerator
 def feq(v1,v2,tol2=1e-4):
   return abs(v1-v2)<=tol2
 
 def ptEq(pt1, pt2, tol=1e-4):
   return feq(pt1.x,pt2.x,tol) and feq(pt1.y,pt2.y,tol) and feq(pt1.z,pt2.z,tol)
 
+# Boost functions are NOT found by doctest, this "fixes" them
```
Comment (greglandrum) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: This is gross/clever. :-)

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 3]
```diff
@@ -106,6 +125,9 @@ def test2DaylightParser(self):
   def test3MDLParsers(self):
     fileN = os.path.join(self.dataDir,'AmideBond.rxn')
     rxna = rdChemReactions.ReactionFromRxnFile(fileN)
+    print ("*"*44)
```
Comment (greglandrum) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: These three calls to print probably aren't really necessary, right?

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py (hunk 2 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: Super gross!

File: Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py [hunk 4]
```diff
@@ -636,6 +658,13 @@ def test23CheckNonProduct(self):
...
+        testFile = os.path.join(RDConfig.RDCodeDir,'Chem','SimpleEnum','test_data','boronic1.rxn')
+        rxn = rdChemReactions.ReactionFromRxnFile(testFile)
+        rxn.Initialize()
+        res = rdChemReactions.PreprocessReaction(rxn)
+        self.assertEquals(res,(0,0,2,1,(((0, 'halogen.bromine.aromatic'),), ((1, 'boronicacid'),))))
+
 
 if __name__ == '__main__':
-  unittest.main()
+  unittest.main(verbosity=True)
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/testReactionWrapper.py]: I'll also turn off verbosity with the print statements.

**Pull request merged**